FIAT/macro.py

```diff
@@ -0,0 +1,186 @@
+
+    def split_topology(self, ref_el, splits=1, variant=None):
+        assert splits == 1
+        sd = ref_el.get_spatial_dimension()
+        top = ref_el.get_topology()
+        new_topology = copy.deepcopy(top)
+        new_topology[sd] = {}
+
+        barycenter = ref_el.make_points(sd, 0, sd+1)
+        old_verts = ref_el.get_vertices()
+        new_verts = old_verts + tuple(barycenter)
+        new_vert_id = len(old_verts)
+
+        new_topology[0][new_vert_id] = (new_vert_id,)
+        for dim in range(1, sd + 1):
+            offset = len(new_topology[dim])
+            for entity, ids in top[dim-1].items():
+                new_topology[dim][offset+entity] = ids + (new_vert_id,)
+        return new_verts, new_topology
+
+
+class IsoSplit(SplitSimplicialComplex):
+
+    def split_topology(self, ref_el, splits=1, variant=None):
+        depth = splits + 1
+        sd = ref_el.get_spatial_dimension()
+        old_verts = ref_el.get_vertices()
+        new_verts = make_lattice(old_verts, depth, variant=variant)
+
+        new_topology = {}
+        new_topology[0] = {i: (i,) for i in range(len(new_verts))}
+        new_topology[1] = {}
+
+        # Loop through vertex pairs
+        # Edges are oriented from low vertex id to high vertex id to avoid duplicates
+        # Place a new edge when the two lattice multiindices are at Manhattan distance < 3,
+        # this connects the midpoints of edges within a face
+        # Only include diagonal edges that are parallel to the simplex edges,
+        # we take the diagonal that goes through vertices at the same depth
+        cur = 0
+        distance = lambda x, y: sum(abs(b-a) for a, b in zip(x, y))
+        for j, v1 in enumerate(lattice_iter(0, depth+1, sd)):
+            for i, v0 in enumerate(lattice_iter(0, depth+1, sd)):
+                if i < j and distance(v0, v1) < 3 and sum(v1) - sum(v0) <= 1:
+                    new_topology[1][cur] = (i, j)
+                    cur = cur + 1
+        if sd == 3:
+            # Cut the octahedron
+            # FIXME do this more generically
+            assert splits == 1
+            new_topology[1][cur] = (1, 8)
+
+        # Get an adjacency list for each vertex
+        edges = new_topology[1].values()
+        adjacency = {v: set(chain.from_iterable(verts for verts in edges if v in verts))
+                     for v in new_topology[0]}
+
+        # Complete the higher dimensional facets by appending a vertex
+        # adjacent to the vertices of codimension 1 facets
+        for dim in range(2, sd+1):
+            entities = []
+            for entity in new_topology[dim-1]:
+                facet = new_topology[dim-1][entity]
+                for v in range(min(facet)):
+                    if set(facet) < adjacency[v]:
+                        entities.append((v,) + facet)
+            new_topology[dim] = dict(enumerate(entities))
+        return new_verts, new_topology
+
+
+class MacroQuadratureRule(QuadratureRule):
```

Comment:
Doc string?

Reply:
Also, we need the constructor (or another class) to somehow support being able to do macro-quadrature over a parent facet